File: Templates/Examples/File Templates/Viper Module.xctemplate/ExampleModuleAssembly.m
//
//  ExampleModuleAssembly.m
//  Проект:   TemplateTester
//
//  Модуль:   Example
//  Описание: Отвечает за демонстрацию возможностей шаблона
//
//  Создан Andrey Zarembo-Godzyatsky  30/07/15
//  Rambler&Co 2015
//

#import "ExampleModuleAssembly.h"
#import "ExampleViewController.h"
#import "ExampleInteractor.h"
#import "ExamplePresenter.h"
#import "ExampleRouter.h"

@interface  ExampleModuleAssembly()

@end

@implementation  ExampleModuleAssembly

- (ExampleViewController *)viewExample {

    return [TyphoonDefinition withClass:[ExampleViewController class]
                          configuration:^(TyphoonDefinition *definition) {
                            [definition injectProperty:@selector(output) 
                                                  with:[self presenterExample]];
             }];
}

- (ExampleInteractor *)interactorExample {

    return [TyphoonDefinition withClass:[ExampleInteractor class]
                          configuration:^(TyphoonDefinition *definition) {
                            [definition injectProperty:@selector(output) 
                                                  with:[self presenterExample]];
             }];
}

- (ExamplePresenter *)presenterExample {

    return [TyphoonDefinition withClass:[ExamplePresenter class]
                          configuration:^(TyphoonDefinition *definition) {
                            [definition injectProperty:@selector(view) 
                                                  with:[self viewExample]];                                            
                            [definition injectProperty:@selector(interactor) 
                                                  with:[self interactorExample]];
                            [definition injectProperty:@selector(router) 
                                                  with:[self routerExample]];
            }];
}

- (ExampleRouter *)routerExample {
    return [TyphoonDefinition withClass:[ExampleRouter class]];
}

@end

File: Templates/Examples/File Templates/Viper Module Swift.xctemplate/ExampleModuleAssembly.swift
//
//  ExampleModuleAssembly.swift
//  Проект:   TemplateTester
//
//  Модуль:   Example
//  Описание: Отвечает за демонстрацию возможностей шаблона
//
//  Создан Andrey Zarembo-Godzyatsky  30/07/15
//  Rambler&Co 2015
//

import UIKit
import Typhoon

public class ExampleModuleAssembly: TyphoonAssembly {

	enum ViperSEL: Selector {
        case view = "view"
        case interactor = "interactor"
        case router = "router"
        case output = "output"
    }

	public dynamic func viewExample() -> AnyObject {
		return TyphoonDefinition.withClass(ExampleView.self) {
			(definition) in 
			
			/*
			//use for init from nib
			definition.useInitializer("initWithNibName:bundle:") {
				(initializer) in

				initializer.injectParameterWith(TyphoonConfig("//XibName//"))
				initializer.injectParameterWith(NSBundle.mainBundle())
			}*/

			definition.injectProperty(ViperSEL.output.rawValue, with: self.presenterExample())
		}
	}


	public dynamic func interactorExample() -> AnyObject {
		return TyphoonDefinition.withClass(ExampleInteractor.self){
            (definition) in
            
            definition.injectProperty(ViperSEL.output.rawValue, with: self.presenterExample())
        }
	}


	public dynamic func presenterExample() -> AnyObject {
		return TyphoonDefinition.withClass(ExamplePresenter.self){
            (definition) in
            
            definition.injectProperty(ViperSEL.view.rawValue, with: self.viewExample())
            definition.injectProperty(ViperSEL.interactor.rawValue, with: self.interactorExample())
            definition.injectProperty(ViperSEL.router.rawValue, with: self.routerExample())
        }
	}


	public dynamic func routerExample() -> AnyObject {
		return TyphoonDefinition.withClass(ExampleRouter.self){
            (definition) in
           
        }
	}
}